ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/OnchainPrivacyController.java

```diff
@@ -0,0 +1,334 @@
+      if (rlpInput.nextSize() > 0) {
+        return Optional.of(
+            new PrivacyGroup(
+                privacyGroupId, PrivacyGroup.Type.ONCHAIN, "", "", decodeList(rlpInput.raw())));
+      } else {
+        return Optional.empty();
+      }
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  @Override
+  public PrivacyGroup[] findPrivacyGroupByMembers(
+      final List<String> addresses, final String privacyUserId) {
+    final ArrayList<PrivacyGroup> privacyGroups = new ArrayList<>();
+    final PrivacyGroupHeadBlockMap privacyGroupHeadBlockMap =
+        privateStateStorage
+            .getPrivacyGroupHeadBlockMap(blockchain.getChainHeadHash())
+            .orElse(PrivacyGroupHeadBlockMap.empty());
+    privacyGroupHeadBlockMap
+        .keySet()
+        .forEach(
+            c -> {
+              final Optional<PrivacyGroup> maybePrivacyGroup =
+                  findPrivacyGroupByGroupId(c.toBase64String(), privacyUserId);
+              if (maybePrivacyGroup.isPresent()
+                  && maybePrivacyGroup.get().getMembers().containsAll(addresses)) {
+                privacyGroups.add(maybePrivacyGroup.get());
+              }
+            });
+    return privacyGroups.toArray(new PrivacyGroup[0]);
+  }
+
+  @Override
+  public PrivacyGroup createPrivacyGroup(
+      final List<String> addresses,
+      final String name,
+      final String description,
+      final String privacyUserId) {
+    throw new PrivacyConfigurationNotSupportedException(
+        "Method not supported when using onchain privacy");
+  }
+
+  @Override
+  public String deletePrivacyGroup(final String privacyGroupId, final String privacyUserId) {
+    throw new PrivacyConfigurationNotSupportedException(
+        "Method not supported when using onchain privacy");
+  }
+
+  @Override
+  public void verifyPrivacyGroupContainsPrivacyUserId(
+      final String privacyGroupId, final String privacyUserId) {
+    verifyPrivacyGroupContainsPrivacyUserId(privacyGroupId, privacyUserId, Optional.empty());
+  }
+
+  @Override
+  public void verifyPrivacyGroupContainsPrivacyUserId(
+      final String privacyGroupId, final String privacyUserId, final Optional<Long> blockNumber) {
+    final Optional<PrivacyGroup> maybePrivacyGroup =
+        onchainPrivacyGroupContract.getPrivacyGroupByIdAndBlockNumber(privacyGroupId, blockNumber);
+    // IF the group exists, check member
+    // ELSE member is valid if the group doesn't exist yet - this is normal for onchain privacy
+    // groups
+    maybePrivacyGroup.ifPresent(
+        (group) -> {
+          if (!group.getMembers().contains(privacyUserId)) {
+            throw new MultiTenancyValidationException(
+                "Privacy group must contain the enclave public key");
+          }
+        });
+  }
+
+  private List<String> getParticipantsFromParameter(final Bytes input) {
+    final List<String> participants = new ArrayList<>();
```

Comment:
Can't we reuse the OnchainUtils here?

Reply:
Didn't know how to nicely use that class here ...

Reply:
Done